backend/tests/test_auth_api.py

```diff
@@ -385,12 +385,61 @@ def test_require_authentication(self, method, url, client: TestClient):
         assert response.status_code == 401
 
     @pytest.mark.parametrize("method, url", authorized)
-    def test_require_authorization(self, method, url, dbsession: Session, testuser: User,
-                                   authenticated_client: TestClient):
-        # The testuser is a superuser. Remove permissions first.
-        dbsession.query(Permission).filter(Permission.recipient_type == RecipientType.USER,
-                                           Permission.recipient_id == testuser.id).delete()
-
+    def test_require_authorization(self, method, url, dbsession: Session,
+                                   unauthorized_testuser: User, authenticated_client: TestClient):
```

Comment:
This test has only been refactored, bu it didn't catch the original bug before, so wouldn't one or a few new tests be in order?

Reply:
Yes, that's what the new `test_review_change_request_breakout_*` are for :slightly_smiling_face: 

We don't want an unauthorized user to change any entity data, but they should be allowed to submit a request

Reply:
But we have not covered the case of an unauthorized user accepting the request. Let me add that...

Reply:
That's great, but my point was that apparently, if I understand the issue correctly, authenticated, but unauthorized users were able to make any kind of change they wanted – directly, i.e. w/o a change request.

Is that correct?

Because if so, then why wasn't *that* caught by any test?

The new tests merely check that such users can create change _requests_ (which themselves don't change anything directly), but isn't there also a way to make changes w/o requests? And if so, where is there a test that unauthorized users cannot make such changes?

Reply:
> why wasn't that caught by any test?

Because the existing test (intended to verify permissions on direct entity changed) checked against the wrong schema (i.e. the one which required reviews). So, the test incorrectly checked that unauthorized users were not allowed to create a change request. By changing to the schema with an unset `reviewable` flag, the test checks what is actually intended: Unauthorized users are not allowed to change an entity.

The new tests cover the other way of doing things: Unauthorized users may create change requests, when the `reviewable` flag is set, but not apply them to the entity.

Does that clarify things?

Reply:
> Because the existing test […]

i.e. `test_require_authorization`, right?

> […] checked against the wrong schema (i.e. the one which required reviews). So, the test incorrectly checked that unauthorized users were not allowed to create a change request. By changing to the schema with an unset `reviewable` flag, the test checks what is actually intended: Unauthorized users are not allowed to change an entity.

This implies that the existing **test** should have been _functionally_ changed in some way (to test the right schema), but it _hasn't!_ The only thing that has changed is that some setup code has been moved into a fixture. Great, but that's only a refactoring.

The test used to pass, when the bug was present, and it hasn't functionally changed.

Why should I now have confidence in a test that previously did not fail in the presence of the exact kind of bug it's meant to detect?

If I were to reintroduce the same bug again, would `test_require_authorization` now fail?

If so, I don't understand why.

If not, then there's still at least one test missing.

I'm open to having a call about this, but ATM it's super unclear.

> The new tests cover the other way of doing things: […]

That part is clear to me and well done for adding those tests.

Reply:
Yes, we are talking about `test_require_authorization`, which is also a parameterized test. In the parameter definition the schema was changed from `unperson` to `person`.

If you were to reintroduce the bug, the testsuite would fail as expected.

If you like, give me a ring...

Reply:
Ah... :bulb: the parametrisation changed!

I meant to ask about that change from `unperson` to `person` and how it ties in, but then I forgot.

Now I see it and it makes sense.